### PR TITLE
providing sig_figs option; fixes #300

### LINKED ISSUE #300: Provide sig_figs options to cmstanMCMC.summary
The summary method executes cmstan's stansummary, and it would be good to pass `sig_figs` option to the summary method. This would make unit testing easier for developers building interfaces to cmstanpy since they can make sure their computations (e.g. their own summary output) line up with what cmdstan does.

https://github.com/stan-dev/cmdstanpy/blob/develop/cmdstanpy/stanfit.py#L616

### PR DESCRIPTION
**Pull request by imadmali**
#### Submission Checklist

- [x] Run unit tests
- [x] Declare copyright holder and open-source license: see below

#### Summary

CmdStan 2.25 introduced output flag "sig_figs", 

- allow this argument for all CmdStan methods (providing CmdStan version is >= 2.25
- add option "sig_figs" to summary method

Unit tests check that draws and summary values have at least the requested amount of precision.

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company): Imad Ali

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)

### CODE REVIEW

**Comment by mitzimorris** [comment]
damn - problems - we need to think this one through better

arg `sig_figs` needs to be specified as a sampler arg in order for the sampler output csv file to capture enough precision.

big  problem with `bin/stansummary` - it will be able to crunch csv files with any amount of precision, but right now, it's got max output sig_figs capped at 10.  this is totally trivial to change on the CmdStan side.

**Comment by mitzimorris** [comment]
just filed https://github.com/stan-dev/cmdstan/pull/937 - this will allow us to request max allowed precision in summary estimates.

to make this work seamlessly, we need to capture `sig_figs` argument from Stan csv header and make that value the max allowable value of `sig_figs` for `summary` function.

**Comment by mitzimorris** [comment]
CmdStan PR got merged - @imadmali - do you want to keep working on this?

**Comment by imadmali** [comment]
@mitzimorris you can go ahead and take over if you like 👍 (it'll take me a while to get up to speed on the cmdstan changes).

**Comment by mitzimorris** [comment]
threaded everything through - hit a snag w/r/t `summary` method - CmdStan PR https://github.com/stan-dev/cmdstan/pull/938 will fix this, so as soon as this goes in, we'll be good - at least on CmdStan `develop` branch.
maybe not for release 2.25.

added arg `sig_figs` to all estimation methods, so right now, users will be able to get extra precision as needed; to calculate stats on a parameter, you can work directly against the drawset.

**Comment by mitzimorris** [comment]
@imadmali and/or @ahartikainen - this is ready for code review - many thanks!

File: cmdstanpy/cmdstan_args.py [hunk 1]
```diff
@@ -619,6 +621,28 @@ def validate(self) -> None:
...
+                or self.sig_figs < 1
+                or self.sig_figs > 18
+            ):
+                raise ValueError(
+                    'sig_figs must be an integer between 1 and 18,'
+                    ' found {}'.format(self.sig_figs)
+                )
+            # sig_figs not valid before version 2.25; allow when version is n/a
+            version = os.path.basename(cmdstan_path())
+            if version.startswith('cmdstan-') and version[8].isdigit():
```
Comment (ahartikainen) [inline cmdstanpy/cmdstan_args.py:636]: Is there a version file somewhere or something we can call with subprocess and parse?

File: cmdstanpy/stanfit.py [hunk 2]
```diff
@@ -636,17 +637,24 @@ def _assemble_draws(self) -> None:
...
+    ) -> pd.DataFrame:
         """
         Run cmdstan/bin/stansummary over all output csv files.
         Echo stansummary stdout/stderr to console.
         Assemble csv tempfile contents into pandasDataFrame.
 
         :param percentiles: Ordered non-empty list of percentiles to report.
             Must be integers from (1, 99), inclusive.
+        :param sig_Figs: Number of significant figures to report.
+            Must be an integer between 1 and 10, inclusive.
```
Comment (ahartikainen) [inline cmdstanpy/stanfit.py:651]: 10 or 18

File: cmdstanpy/stanfit.py [hunk 3]
```diff
@@ -636,17 +637,24 @@ def _assemble_draws(self) -> None:
...
+        self, percentiles: List[int] = None, sig_figs: int = None
+    ) -> pd.DataFrame:
         """
         Run cmdstan/bin/stansummary over all output csv files.
         Echo stansummary stdout/stderr to console.
         Assemble csv tempfile contents into pandasDataFrame.
 
         :param percentiles: Ordered non-empty list of percentiles to report.
             Must be integers from (1, 99), inclusive.
+        :param sig_Figs: Number of significant figures to report.
```
Comment (ahartikainen) [inline cmdstanpy/stanfit.py:650]: sig_figs or sig_Figs

**Review by ahartikainen — APPROVED** [review]

File: cmdstanpy/cmdstan_args.py (hunk 1 above)
Comment (mitzimorris) [reply cmdstanpy/cmdstan_args.py:636]: yes, we can either grep "CMDSTAN_VERSION" out of the makefile, or call `make print-CMDSTAN_VERSION`

File: cmdstanpy/stanfit.py (hunk 2 above)
Comment (mitzimorris) [reply cmdstanpy/stanfit.py:651]: good catch - 18

File: cmdstanpy/stanfit.py (hunk 3 above)
Comment (mitzimorris) [reply cmdstanpy/stanfit.py:650]: def sig_Figs - thanks

**Comment by mitzimorris** [comment]
@ahartikainen many thanks for the code review - your question made me realize that we need utility function `cmdstan_version_at` so that CmdStanPy can handle new CmdStan features without breaking things for folks who are still using older versions of CmdStan.  code and tests are a whole lot cleaner.

**Pull request merged**